File: src/main/scala/org/rutiger/theatre/actors/exercise/third/BattleScribe.scala
package org.rutiger.theatre.actors.exercise.third

import akka.actor.{Actor, ActorLogging, DeadLetter, Props}
import akka.event.Logging.Info

class BattleScribe extends Actor with ActorLogging {

  val eventStream = context.system.eventStream

  def receive = {
    case DeadLetter(message, sender, receiver) => {
      val msg = s"Cannot deliver message [${message.getClass.getName}] from $sender to $receiver"
      log.warning("BATTLE LOG => {}", msg)
      eventStream.publish(Info(receiver.path.toString, receiver.getClass, msg))
    }
  }

  override def preStart(): Unit = eventStream.subscribe(self, classOf[DeadLetter])

  override def postStop(): Unit = eventStream.unsubscribe(self)
}


object BattleScribe {
  def apply: Props = Props[BattleScribe]
}

File: src/main/java/org/rutiger/theatre/actors/exercise/third/BattleScribe.java
package org.rutiger.theatre.actors.exercise.third;

import akka.actor.AbstractActor;
import akka.actor.DeadLetter;
import akka.actor.Props;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;
import org.rutiger.theatre.Messages;


public class BattleScribe extends AbstractActor {

    private LoggingAdapter log = Logging.getLogger(getContext().system(), this);

    //Receive a dead letter message and process it.
    @Override
    public Receive createReceive() {
        return receiveBuilder()
                .match(DeadLetter.class, dlm -> {
                    String message = String.format("Cannot deliver message '%s' from %s to %s",
                            dlm.getClass().getName(), dlm.sender().path(), dlm.recipient().path());
                    log.info("Battle LOG: {}", message);
                    getContext().getSystem().eventStream().publish(message);
                }).build();
    }

    @Override
    public void preStart() throws Exception {
        getContext().getSystem().eventStream().subscribe(getSelf(), DeadLetter.class);
    }

    @Override
    public void postStop() throws Exception {
        getContext().getSystem().eventStream().unsubscribe(getSelf());
    }

    public static Props props() {
        return Props.create(BattleScribe.class);
    }
}
